lib/plugins/paragon-webpack-plugin/ParagonWebpackPlugin.js

```diff
@@ -0,0 +1,139 @@
+const { Compilation, sources } = require('webpack');
+const fs = require('fs');
+const path = require('path');
+const {
+  getParagonVersion,
+  getParagonThemeCss,
+} = require('../../../config/data/paragonUtils');
+const {
+  injectMetadataIntoDocument,
+  getParagonStylesheetUrls,
+  injectParagonCoreStylesheets,
+  injectParagonThemeVariantStylesheets,
+} = require('./utils');
+
+// Resolve configuration `env.config`.
+const envConfigFile = path.resolve(process.cwd(), 'env.config.js');
+
+// Get Paragon and brand versions / CSS files from disk.
+const paragonVersion = getParagonVersion(process.cwd());
+const paragonThemeCss = getParagonThemeCss(process.cwd());
+const brandVersion = getParagonVersion(process.cwd(), { isBrandOverride: true });
+const brandThemeCss = getParagonThemeCss(process.cwd(), { isBrandOverride: true });
+
+/**
+ * 1. Injects `PARAGON_THEME` global variable into the HTML document during the Webpack compilation process.
+ * 2. Injects `<link rel="preload" as="style">` element(s) for the Paragon and brand CSS into the HTML document.
+ */
+class ParagonWebpackPlugin {
+  constructor({ processAssetsHandlers = [] } = {}) {
+    this.pluginName = 'ParagonWebpackPlugin';
+    this.paragonThemeUrlsConfig = {};
+    this.paragonMetadata = {};
+
+    // List of handlers to be executed after processing assets during the Webpack compilation.
+    this.processAssetsHandlers = [
+      this.resolveParagonThemeUrlsFromConfig,
+      this.injectParagonMetadataIntoDocument,
+      this.injectParagonStylesheetsIntoDocument,
+      ...processAssetsHandlers,
+    ].map(handler => handler.bind(this));
+  }
+
+  /**
+   * Resolves the MFE configuration for ``PARAGON_THEME_URLS`` in the following priority order:
+   *
+   * 1. ``envConfig.PARAGON_THEME_URLS`` (JS based)
+   * 3. ``process.env.PARAGON_THEME_URLS`` (JSON based)
```

Comment:
I feel like having `PARAGON_THEME_URLS` work in an `env.config.js` file but not an `env.config.jsx` file is more confusing than just saying `PARAGON_THEME_URLS` *must* be set as an environment variable.

I'd love to hear what others think about this!

Reply:
To be clear - this file _cannot_ import or use `env.config` with any file extension, and we must use `process.env` to get this value.  The env.config file is _runtime_ configuration and is expected to import runtime files that will require transpilation (i.e., jsx/tsx) and depend on their own runtime dependencies.  That code can't meaningfully pollute the build-time configuration, and having multiple files named env.config with different extensions would be seriously confusing for folks.  

At a later date we may decide to create a new type of build-only configuration file, but we're just not there at the moment.  For now, we just need to use process.env either via command line environment variables or the `.env` files, and it'll come in as a string we need to `JSON.parse`.

Reply:
Furthermore, worse, would be for build-time code and build-time dependencies to find their way into env.config... that would mean build-time code gets bundled with the app for a browser environment which is a huge no-no and would definitely break.

Reply:
+1. Agreed, we should only be relying on `process.env.PARAGON_THEME_URLS` at this point, with `JSON.parse` for the build-time vs. runtime configuration that @davidjoy is alluding to above (across both this frontend-build PR _and_ within frontend-platform).

Reply:
On the runtime side for frontend-platform's use, by all means copy PARAGON_THEME_URLS into env.config and format it nicely as an actual object instead of a string.  But for now it does need to be duplicated, unfortunately, because the two types of config can't mingle.

Reply:
I gave a couple of thumbs up above, but better to be explicit: I agree that for now, `process.env.PARAGON_THEME_URLS` should be what is used, and only that.

I also like the idea of eventually having a separate build-only config file, thus doing away with any potential for confusion or breakage.  (While at the same time finally removing support for the .env files, hopefully.)

Reply:
> I also like the idea of eventually having a separate build-only config file, thus doing away with any potential for confusion or breakage. (While at the same time finally removing support for the .env files, hopefully.)

### :100: 

That lines up perfectly with the feelings expressed in [the slack thread](https://openedx.slack.com/archives/C04BM6YC7A6/p1723069071930839?thread_ts=1723049201.030959&cid=C04BM6YC7A6).

Reply:
Done 😁